agr_java_core/src/main/java/org/alliancegenome/neo4j/repository/AlleleRepository.java

```diff
@@ -58,28 +63,75 @@ public Allele getAllele(String primaryKey) {
         return list;
     }
 
-    public Set<Allele> getAllAlleles() {
-        HashMap<String, String> map = new HashMap<>();
+    private Map<String, Transcript> transcriptMap;
 
+    public Map<String, Transcript> getTranscriptWithExonInfo() {
+        if (MapUtils.isNotEmpty(transcriptMap))
+            return transcriptMap;
         String query = "";
-        query += " MATCH p1=(:Species)<-[:FROM_SPECIES]-(a:Allele) ";
-        //query += " where g.primaryKey = 'FB:FBgn0002121' AND a.primaryKey = 'FB:FBal0051412' ";
-        //query += " where a.primaryKey in ['MGI:3795217','MGI:3712283','MGI:3843784','MGI:2158359'] ";
-        query += " OPTIONAL MATCH gene=(a:Allele)-[:IS_ALLELE_OF]->(g:Gene)-[:FROM_SPECIES]-(q:Species)";
-        query += " OPTIONAL MATCH vari=(a:Allele)<-[:VARIATION]-(variant:Variant)--(soTerm:SOTerm)";
-        query += " OPTIONAL MATCH consequence=(:GeneLevelConsequence)<-[:ASSOCIATION]-(variant:Variant)";
-        query += " OPTIONAL MATCH loc=(variant:Variant)-[:ASSOCIATION]->(:GenomicLocation)-[:ASSOCIATION]->(:Chromosome)";
-        query += " OPTIONAL MATCH variantPub=(variant:Variant)-[:ASSOCIATION]->(:Publication)";
-        query += " OPTIONAL MATCH p2=(a:Allele)-[:ALSO_KNOWN_AS]->(synonym:Synonym)";
-        query += " OPTIONAL MATCH crossRef=(a:Allele)-[:CROSS_REFERENCE]->(c:CrossReference)";
+        // get Transcript - Exon relationships
+        query += " MATCH p1=(t:Transcript)-[:ASSOCIATION]->(:GenomicLocation)-[:ASSOCIATION]->(:Chromosome) ";
+        query += " OPTIONAL MATCH p2=(:GenomicLocation)<-[:ASSOCIATION]-(:Exon)-[:EXON]->(t:Transcript) ";
+        query += " RETURN p1, p2";
+        Iterable<Transcript> transcriptExonsIter = query(Transcript.class, query);
+        log.info("Number of Transcript/Exon relationships: " + String.format("%,d", (int) StreamSupport.stream(transcriptExonsIter.spliterator(), false).count()));
+        transcriptMap = StreamSupport.stream(transcriptExonsIter.spliterator(), false)
+                .collect(Collectors.toSet())
+                .stream()
+                .collect(Collectors.toMap(Transcript::getPrimaryKey, transcript -> transcript));
+        return transcriptMap;
+    }
+
+    public Set<Allele> getAllAlleles() {
+        String query = "";
+        // allele-only (no variants)
+        query += " MATCH p1=(:Species)<-[:FROM_SPECIES]-(a:Allele)-[:IS_ALLELE_OF]->(g:Gene)-[:FROM_SPECIES]-(q:Species) ";
+        query += "where not exists ((a)<-[:VARIATION]-(:Variant)) ";
```

Comment:
@cmpich does this `(a)` need to be `(a:Allele)`?

Reply:
I found this syntax, though, I am sure you could also add the :Allele to it.

Reply:
Ya this will effect performance as in (a) would be ALL nodes connected to `[:VARIATION]`